syncd/syncd_pfc_watchdog.h

```diff
@@ -0,0 +1,65 @@
+#ifndef PFC_WATCHDOG_H
+#define PFC_WATCHDOG_H
+
+extern "C" {
+#include "sai.h"
+}
+
+#include <atomic>
+#include <vector>
+#include <condition_variable>
+#include "swss/table.h"
+
+class PfcWatchdog
+{
+    public:
+        static void setPortCounterList(
+                _In_ sai_object_id_t queueVid,
+                _In_ sai_object_id_t queueId,
+                _In_ const std::vector<sai_port_stat_t> &counterIds);
+        static void setQueueCounterList(
+                _In_ sai_object_id_t queueVid,
+                _In_ sai_object_id_t queueId,
+                _In_ const std::vector<sai_queue_stat_t> &counterIds);
+        static void removeQueue(
+                _In_ sai_object_id_t queueVid);
+
+        PfcWatchdog(
+                _In_ const PfcWatchdog&) = delete;
+        ~PfcWatchdog(void);
+
+    private:
+        struct PfcCounterIds
+        {
+            PfcCounterIds(
+                    _In_ sai_object_id_t queue,
+                    _In_ sai_object_id_t port,
+                    _In_ const std::vector<sai_port_stat_t> &portIds,
+                    _In_ const std::vector<sai_queue_stat_t> &queueIds);
+
+            sai_object_id_t queueId;
+            sai_object_id_t portId;
+            std::vector<sai_port_stat_t> portCounterIds;
+            std::vector<sai_queue_stat_t> queueCounterIds;
```

Comment:
it is better to organize your counterIDmap based on the port id for the all port counters.

In this case, you can query all the port counters for one port together. later, you can write to the counter db indexed by queue id, since it will be easier for the lua script.

the pfccounterids has following structure

portqueuecounterids
{
portid;  // valid only for for portcounterid
std::vector<tuple<queueid, sai_port_stat_t>> portCounterIds; // valid only for portcounter id
queueid; // valid only for queuecounterid.
vector<sai_queue_stat_t> queueids // validonly for queuecounter id
}

Reply:
Agree. I will split subscription for queue counters and port counters so that port counters will be queried only once.